File: Java_Version/src/Player.java
import java.util.List;
import java.util.Random;
import java.util.ArrayList;

public class Player {

	private int shots = 0;
	private int missed_shots = 0;
	private int hit_shots = 0;
	private int repeated_shots = 0;

	private String name;
	
	public Player(String name) {
		this.name = name;
	}

	public void placeAllShips(Board br) {

		orientation orien;
		int ori, xi = 0, yi = 0;
		List< Integer > crds = new ArrayList< Integer >(2);
		crds.add(0);
		crds.add(1);
		
		boolean plc;

		Random rand = new Random();
		
		for (int i = 1; i <= 5; i++) {
			
			ori = 0;
			if (i == 1) {
				do{
					
					ori = rand.nextInt(2) + 0;
					if (ori == 1)
						orien = orientation.vertical;
					else
						orien = orientation.horizontal;

					xi = rand.nextInt(br.getLength()) + 0;
					//System.out.print("AAA"+xi+"AAA");
					yi = rand.nextInt(br.getLength()) + 0;
					crds.set(0,xi);
					crds.set(1,yi);

					Ship cr = new Carrier(br.getBrd().get(xi).get(yi), orien);
					plc = placeShip(cr, br.getBrd().get(xi).get(yi), cr.getOrient(), br);
				} while (plc==false);
			}
			else if (i == 2) {
				do {
					
					ori = rand.nextInt(2) + 0;
					if (ori == 1)
						orien = orientation.vertical;
					else
						orien = orientation.horizontal;

					xi = rand.nextInt(br.getLength()) + 0;
					yi = rand.nextInt(br.getLength()) + 0;
					
					crds.set(0,xi);
					crds.set(1,yi);

					BattleS bs = new BattleS(br.getBrd().get(xi).get(yi), orien);
					plc = placeShip(bs, br.getBrd().get(xi).get(yi), bs.getOrient(), br);
				} while (plc==false);
			}
			else if (i == 3) {
				do {

					ori = rand.nextInt(2) + 0;
					if (ori == 1)
						orien = orientation.vertical;
					else
						orien = orientation.horizontal;

					xi = rand.nextInt(br.getLength()) + 0;
					yi = rand.nextInt(br.getLength()) + 0;
					crds.set(0,xi);
					crds.set(1,yi);

					Ship cu = new Cruiser(br.getBrd().get(xi).get(yi), orien);
					plc = placeShip(cu, br.getBrd().get(xi).get(yi), cu.getOrient(), br);
				} while (plc==false);
			}
			else if (i == 4) {
				do {
					
					ori = rand.nextInt(2) + 0;
					if (ori == 1)
						orien = orientation.vertical;
					else
						orien = orientation.horizontal;

					xi = rand.nextInt(br.getLength()) + 0;
					yi = rand.nextInt(br.getLength()) + 0;
					crds.set(0,xi);
					crds.set(1,yi);

					Ship ds = new Destroyer(br.getBrd().get(xi).get(yi), orien);
					plc = placeShip(ds, br.getBrd().get(xi).get(yi), ds.getOrient(), br);
				} while (plc==false);
			}
			else if (i == 5) {
				do {

					ori = rand.nextInt(2) + 0;
					if (ori == 1)
						orien = orientation.vertical;
					else
						orien = orientation.horizontal;

					xi = rand.nextInt(br.getLength()) + 0;
					yi = rand.nextInt(br.getLength()) + 0;
					crds.set(0,xi);
					crds.set(1,yi);

					Ship sm = new Submarine(br.getBrd().get(xi).get(yi), orien);
					plc = placeShip(sm, br.getBrd().get(xi).get(yi), sm.getOrient(), br);
				} while (plc==false);
			}
		}
	}
	
	
	public boolean placeShip( Ship sh, Tile coord, orientation ornt, Board br) {
	
		//System.out.println("aaa21aaa");
		boolean res = sh.placeShip( coord, ornt, br);
		//System.out.println( "fffff" );
	
		return res;
	}

	public void statsInit() {
		this.hit_shots = 0;
		this.missed_shots = 0;
		this.repeated_shots = 0;
		this.shots = 0;
	}

	public void getStats() {
		System.out.println("Player "+ this.name +" has: \n\t "+ this.hit_shots +" / "+ this.shots +" hit shots !");
		System.out.println("\n\t "+ this.missed_shots + " / " + this.shots + " missed shots !");
		System.out.println(" \n\t and " + this.repeated_shots + " / " + this.shots + " repeated shots ! \n" );
	}
	
	public boolean fire( Board brd, List<Integer> coord) {

		boolean check = false;
		
		sea_type target = brd.getBrd().get( coord.get(0) ).get( coord.get(1) ).getType();
		// hit
		if (target == sea_type.ship) {
			this.shots++;
			this.hit_shots++;
			brd.setBrd( coord.get(0), coord.get(1), sea_type.hit);
			System.out.println(" * Hit ! * ");
			check = true;
		}
		// already hit
		else if (target == sea_type.hit) {
			this.shots++;
			this.repeated_shots++;
			brd.setBrd( coord.get(0), coord.get(1), sea_type.hit); //brd[coord[0]][coord[1]].setType(hit);
			System.out.println(" * Already Hit ! * ");
			check = true;
		}
		// already miss
		else if (target == sea_type.miss) {
			this.shots++;
			this.repeated_shots++;
			brd.setBrd(coord.get(0), coord.get(1), sea_type.miss); //brd[coord[0]][coord[1]].setType(hit);
			System.out.println(" * Already Miss ! * ");
		}
		// miss
		else if (target == sea_type.sea) {
			this.shots++;
			this.missed_shots++;
			brd.setBrd(coord.get(0), coord.get(1), sea_type.miss); //brd[coord[0]][coord[1]].setType(miss);
			System.out.println(" * Miss ! * ");
		}
		
		return check;
	}
	
	
}

File: Cpp_Version/player.cpp
#include "stdafx.h"
#include "player.h"
#include "ship.h"
#include "Carrier.h"
#include "Submarine.h"
#include "Cruiser.h"
#include "Destroyer.h"
#include "Battleship.h"
#include "tile.h"
#include <ctime>
#include "board.h"

void Player::setName(char* s1) {
	this->name = s1;
}

void Player::placeAllShips(Board &br) {

	orientation orien;
	int ori, xi, yi;
	std::vector< int > crds(2);
	Ship *cr, *bs, *cu, *ds, *sm;
	bool plc;

	std::srand(std::time(nullptr));
	for (int i = 1; i <= 5; i++) {
		
		ori = 0;
		if (i == 1) {
			do{
				
				ori = std::rand() / ((RAND_MAX + 1u) / 2);
				if (ori == 1)
					orien = vertical;
				else
					orien = horizontal;


				xi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				yi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				crds[0] = xi;
				crds[1] = yi;

				Ship* cr = new Carrier(br.getBrd()[xi][yi], orien);
				plc = placeShip(cr, br.getBrd()[xi][yi], cr->getOrient(), br);
			} while (plc==false);
		}
		else if (i == 2) {
			do {
				
				ori = std::rand() / ((RAND_MAX + 1u) / 2);
				if (ori == 1)
					orien = vertical;
				else
					orien = horizontal;

				xi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				yi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				crds[0] = xi;
				crds[1] = yi;

				Ship* bs = new Battleship(br.getBrd()[xi][yi], orien);
				plc = placeShip(bs, br.getBrd()[xi][yi], bs->getOrient(), br);
			} while (plc==false);
		}
		else if (i == 3) {
			do {
				
				ori = std::rand() / ((RAND_MAX + 1u) / 2);
				if (ori == 1)
					orien = vertical;
				else
					orien = horizontal;

				xi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				yi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				crds[0] = xi;
				crds[1] = yi;

				Ship* cu = new Cruiser(br.getBrd()[xi][yi], orien);
				plc = placeShip(cu, br.getBrd()[xi][yi], cu->getOrient(), br);
			} while (plc==false);
		}
		else if (i == 4) {
			do {
				
				ori = std::rand() / ((RAND_MAX + 1u) / 2);
				if (ori == 1)
					orien = vertical;
				else
					orien = horizontal;

				xi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				yi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				crds[0] = xi;
				crds[1] = yi;

				Ship* ds = new Destroyer(br.getBrd()[xi][yi], orien);
				plc = placeShip(ds, br.getBrd()[xi][yi], ds->getOrient(), br);
			} while (plc==false);
		}
		else if (i == 5) {
			do {
				
				ori = std::rand() / ((RAND_MAX + 1u) / 2);
				if (ori == 1)
					orien = vertical;
				else
					orien = horizontal;

				xi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				yi = std::rand() / ((RAND_MAX + 1u) / br.getLength());
				crds[0] = xi;
				crds[1] = yi;

				Ship* sm = new Submarine(br.getBrd()[xi][yi], orien);
				plc = placeShip(sm, br.getBrd()[xi][yi], sm->getOrient(), br);
			} while (plc==false);
		}
	}
}

bool Player::placeShip(Ship* sh, Tile coord, orientation ornt, Board &br) {

	//printf("aaa21aaa");
	bool res = sh->placeShip(coord, ornt, br);
	//printf( "fffff" );

	return res;
}

void Player::statsInit() {
	this->hit_shots = 0;
	this->missed_shots = 0;
	this->repeated_shots = 0;
	this->shots = 0;
}

void Player::getStats() {
	printf("Player %s has: \n\t %d / %d hit shots !", this->name, this->hit_shots, this->shots);
	printf("\n\t %d / %d missed shots !", this->missed_shots, this->shots);
	printf(" \n\t and %d / %d repeated shots ! \n", this->repeated_shots, this->shots);
}

bool Player::fire(Board &brd, std::vector<int> coord) {
	
	bool check = false;

	sea_type target = brd.getBrd()[coord[0]][coord[1]].getType();
	// hit
	if (target == ship) {
		brd.setBrd(coord[0], coord[1], hit);
		printf(" * Hit ! * ");
		this->shots++;
		this->hit_shots++;
		check = true;
	}
	// already hit
	else if (target == hit) {
		brd.setBrd( coord[0], coord[1], hit); //brd[coord[0]][coord[1]].setType(hit);
		printf(" * Already Hit ! * ");
		this->repeated_shots++;
		this->shots++;
		check = true;
	}
	// already miss
	else if (target == miss) {
		brd.setBrd(coord[0], coord[1], miss); //brd[coord[0]][coord[1]].setType(hit);
		printf(" * Already Miss ! * ");
		this->repeated_shots++;
		this->shots++;
	}
	// miss
	else if (target == sea) {
		brd.setBrd(coord[0], coord[1], miss); //brd[coord[0]][coord[1]].setType(miss);
		printf(" * Miss ! * ");
		this->missed_shots++;
		this->shots++;
	}

	return check;
}
